File: aufgabe_3_lazyloading/php/LazyMenuProxy.php
<?php

interface MenuInterface {
    public function display(): void;
}

class RealMenuManager implements MenuInterface {
    private string $fileName;

    public function __construct(string $fileName) {
        $this->fileName = $fileName;
        $this->loadPdfFromDisk();
    }

    private function loadPdfFromDisk(): void {
        echo "[RAM] Lade riesige PDF-Datei ({$this->fileName}) in den Arbeitsspeicher... (Verbraucht 50MB)\n";
    }

    public function display(): void {
        echo "[Anzeige] Zeige Speiseplan {$this->fileName} barrierefrei auf dem Bildschirm an.\n";
    }
}

class LazyMenuProxy implements MenuInterface {
    private ?RealMenuManager $realMenuManager = null; // Standardmäßig null
    private string $fileName;

    public function __construct(string $fileName) {
        // TODO: Speichere nur den Dateinamen ab. 
        // Erstelle HIER NOCH NICHT das RealMenuManager-Objekt!
    }

    public function display(): void {
        // TODO: Prüfe, ob $this->realMenuManager gleich null ist.
        // Wenn JA: Erstelle das Objekt jetzt: new RealMenuManager($this->fileName);
        
        // Rufe danach die display() Methode auf dem echten Objekt auf.
    }
}

// --- TEST BEREICH ---
// $menu = new LazyMenuProxy("Speiseplan_Juni_2026.pdf"); 
// echo "Proxy wurde erstellt. Noch ist kein RAM verbraucht...\n";
// $menu->display(); // Erst JETZT wird die Datei geladen und angezeigt!

File: aufgabe_3_lazyloading/python/lazy_menu_proxy.py
from abc import ABC, abstractmethod

class MenuInterface(ABC):
    @abstractmethod
    def display(self):
        pass

class RealMenuManager(MenuInterface):
    def __init__(self, file_name: str):
        self.file_name = file_name
        self._load_pdf_from_disk()

    def _load_pdf_from_disk(self):
        print(f"[RAM] Lade riesige PDF-Datei ({self.file_name}) in den Arbeitsspeicher... (50MB)")

    def display(self):
        print(f"[Anzeige] Zeige Speiseplan {self.file_name} auf dem Bildschirm an.")

class LazyMenuProxy(MenuInterface):
    def __init__(self, file_name: str):
        # TODO: Speichere den Dateinamen. 
        # Setze self.real_menu_manager initial auf None.
        pass

    def display(self):
        # TODO: Prüfe, ob self.real_menu_manager None ist.
        # Wenn JA: Erstelle jetzt die Instanz von RealMenuManager.
        
        # Rufe danach display() auf dem echten Objekt auf.
        pass

# --- TEST BEREICH ---
# menu = LazyMenuProxy("Speiseplan_Juni_2026.pdf")
# print("Proxy erstellt. Kein RAM verbraucht...")
# menu.display() # Erst jetzt wird geladen!
